File: python/BoxChallenge.py
from controller import Robot
from controller import Gyro

#function definitions
'''
Updates angle variable according to angular velocity from gyro
AngleCurrent = anglePast + integral of angular velocity over one timeStep since last updated angle
Should be called every time main loop repeats
'''
def updateGyro():
    global angle
    angle += (timeStep / 1000.0) * (gyro.getValues())[1]

'''
Returns current angle of robot relative to starting angle
Angle does not drop to 0 after exceeding 360
Angle % 360 will yield relative angle with maximum 360
'''
def getAngle():
    return angle * 180 / 3.1415

'''
Sets motors to go forward dist meters
'''
def goForward(leftMotor, rightMotor, dist):
    global state, targetEnc
    if encoders.getValue() == encoders.getValue():  #checks if encoder value is not naN (enocder = naN before motor first moves)
        targetEnc = encoders.getValue() + radPerMeter * dist
    else:
        targetEnc = radPerMeter * dist
    state = forward
    leftMotor.setVelocity(maxVelocity * 0.5)
    rightMotor.setVelocity(maxVelocity * 0.5)

'''
Checks if encoders have not reached the target amount of encoders (targetEnc) while going forward
Stops motors and returns false if reached
Returns true otherwise
'''
def ifForward(leftMotor, rightMotor):
    if encoders.getValue() - targetEnc > -0.01:
        leftMotor.setVelocity(0)
        rightMotor.setVelocity(0)
        return False
    return True

'''
Sets motors to turn
Notes starting angle so ifTurn90() can find degrees turned so far
'''
def doTurn(leftMotor, rightMotor):
    global state, initAngle
    initAngle = round(getAngle() / 90.0) * 90.0     #Saves angle before turn rounded to the neared 90 degrees
    state = turn
    leftMotor.setVelocity(maxVelocity * 0.1)        #10% of maxVelocity or else overshoots - change ifTurn90() to help overshoot
    rightMotor.setVelocity(-maxVelocity * 0.1)

'''
Checks compass angle to see if robot must turn more to complete 90 degree turn
Stops motors and returns false if reached
Returns true otherwise
'''
def ifTurn90():
    if abs(getAngle() - initAngle) >= 86.0:     #Increase 86.0 to help undershoot, decrease for overshoot
        leftMotor.setVelocity(0)
        rightMotor.setVelocity(0)
        return False
    return True

#variable definitions
timeStep = 32           ''' timeStep is the interval in milliseconds between iterations of the
                        main Webots loop  (while myRobot.step(timeStep) != -1)
                        Code inside of main loop should not take longer than timeStep milliseconds '''
targetEnc = 0           ''' target amount of radians for wheel to spin when robot goes forward
                            robot position sensor (encoder counter) cannot be reset, so always add to targetEnc '''
numEdgesTravel = 0      #Increases every time robot travels a side of the box 
angle = 0               #Stores current angle of robot relative to start - updated by updateGyro
initAngle = 0           #Stores starting angle of a turn so ifTurn90() knows when 90 degrees has been turned
radPerMeter = 10.5      #Converts target distance in meters into radians the wheel must spin to travel that distance


state = 0               ''' Webots robot controller require that the main loop (while myRobot.step(timeStep) != -1)
                            run once every timeStep milliseconds -> cannot run commands that take 
                            time sequentially e.g. motor movements (go forward, turn, etc.)
                            Use state variable to store what movement is currently being executed so
                            main loop knows when to change movements and which movement to change to
                            e.g. if state == forward (value of 1, see below), main loop should check if
                            encoder target is reached and begin turning if it is, upon which state changed to turn '''
                            
forward = 1             '''state values for specific motor movements'''
turn = 2                '''e.g. state = 1 if robot is currently moving forward'''

'''
Robot, motors, and all sensors (devices) need to be represented by objects
Use get____ method, passing the name of the device as a parameter, to get objects
Use the objects to call methods related to that device
'''
myRobot = Robot()
leftMotor = myRobot.getMotor('left wheel')
rightMotor = myRobot.getMotor('right wheel')
encoders = leftMotor.getPositionSensor()         '''Position sensor stores radians turned since beginning of the program
                                                    Counter cannot be reset, so targetEnc must always be updated 
                                                    relative to cummulative radians turned'''
gyro = myRobot.getGyro('gyro')                  '''Gyro sensor returns angular velocity in 3 directions, must integrate 
                                                    for true angle updateGyro() function provided to update angle 
                                                    variable to horizontal compass value'''

'''
Initialize motors and sensors
timeStep must be passed to sensor.enable to show how often probing will occur
'''
leftMotor.setPosition(float("inf"))     '''setPosition sets motors to turn a target radians that you pass in'''
rightMotor.setPosition(float("inf"))    '''Motors will not move once target is reached, no matter what velocity is set to
                                            Pass it infinity so motors will never reach target and stop automatically,
                                            so now motors are controlled by only setVelocity.
                                            We can still change the target position by calling setPosition later in the program'''
leftMotor.setVelocity(0.0)              #Sets motors to rotate at certain speed in radians per second (max velocity = 10)
rightMotor.setVelocity(0.0)
gyro.enable(timeStep)
encoders.enable(timeStep)
maxVelocity = leftMotor.getMaxVelocity()

#main code
''' 
Take note of state variable
    Code must check current state and also if it is done with that state
    If so, sends new commands to motor and switches state appropriately
    Necessary to deal with the main loop that must iterate every 32 ms 
'''
state = forward
goForward(leftMotor, rightMotor, 2.0)
while myRobot.step(timeStep) != -1:     #main loop - runs every 32 ms (because timeStep = 32)
    updateGyro()
    if state == forward and not ifForward(leftMotor, rightMotor):   #if robot is finished going forward
        doTurn(leftMotor, rightMotor)
    elif state == turn and not ifTurn90():      #if robot is finished turning
        if numEdgesTravel < 3:
            numEdgesTravel += 1
            goForward(leftMotor, rightMotor, 2.0)
        

File: ErebusSamples/cpp/BoxChallenge.cpp
/*
 * BoxChallenge.cpp:
 *
 * Date: June 8, 2020
 * Author: Victor Hu and Jeffrey Cheng
 * Facilitator: Storming Robots
 * Version: 1.0 initial release
 * Description:
 * Sample code to demonstrate how to navigate the robot in a square pattern using the motors, encoders, and gyro sensor.
 * The purpose of this sample is to serve as a starting point to learn from. It does not exhibit a fully functioning robot.
 *
 * License: You may use this code as you wish, provided you give credit where it's due.
 */

// C++ include statements for input/output and math library
#include <iostream>
#include <Math.h>
using namespace std;

// Webots include statements 
#include <webots/Robot.hpp>
#include <webots/Motor.hpp>
#include <webots/Gyro.hpp>
#include <webots/PositionSensor.hpp>
using namespace webots;

// Instantiate new Robot object
Robot* robot = new Robot();

// Gets pointers to the motor and sensors from the robot
Motor* leftMotor = robot->getMotor("left wheel"), *rightMotor = robot->getMotor("right wheel");
Gyro* gyro = robot->getGyro("gyro");
PositionSensor* leftEnc = leftMotor->getPositionSensor(), *rightEnc = rightMotor->getPositionSensor();

// Get the max velocity the motors are allowed to go at, in rad/s.
const int MAX_VELOCITY = leftMotor->getMaxVelocity();

/*
 * Number of radians the motors must turn to move the robot forward 1 meter.
 * This is found by taking 100cm (1m) and dividing it by pi * 19cm, the 
 * cirumference of the wheel of the pioneer3DX robot found at
 * https://cyberbotics.com/doc/guide/pioneer-3dx which yields wheel revolutions per meter.
 * Then multiply by 2pi to convert to radians per meter.
 */
const double radPerMeter = 10.5;

/* Variable declaration */

// Variable "angle" stores our current angle heading. "initAngle" is used for keeping track of how
// much we have turned.
double angle = 0, initAngle;

// Use an enum called "state" in order to store our state. This state alternates between moving forward
// and turning.
enum State { movingForward, turn } state;

// numEdgesTravel keeps track of how many sides of the box we have currently traveled. Once numEdgesTravel reaches 4,
// then we know that we have completely gone around in a square pattern. targetEnc stores our desired encoder value
// when we want to move forward a certain amount of meters.
int numEdgesTravel = 0, targetEnc = 0;

/*
 * Webots gyros return angular velocity along the 3 axes. To find angle, we need to integrate
 * angular velocity by multiplying by the timeStep. We only care about rotation around the y-axis AKA yaw.
 * The result is stored in the variable "angle" in radians.
 */
void updateGyro(int timeStep) {
    angle += (timeStep / 1000.0) * (gyro->getValues())[1];
}

// Converts the current angle from radians into degrees.
// 180 degrees = pi radians
double getAngle() {
    return angle * 180 / 3.1415;
}

// Makes the robot go forward for a setpoint of "dist" meters.
void goForward(double dist) {

	// Sets the robot state to moving forward
    state = movingForward;

	// Calculates the target encoder value by converting "dist" into radians
	// and adding it onto the current encoder value.
    targetEnc = leftEnc->getValue() + radPerMeter * dist;

	// Sets the left and right motors to move forward at half the max velocity
    leftMotor->setVelocity(MAX_VELOCITY * .5);
    rightMotor->setVelocity(MAX_VELOCITY * .5);
}

// Makes the robot do a clockwise 90 degrees point turn
void doTurn() {

	// Sets the robot state to turning
    state = turn;

	// Updates initAngle to be the current angle (in degrees).
    initAngle = getAngle();

	// Sets the left motor to move forward at 10% of max velocity.
	// Sets the right motor to move backward at 10% of max velocity.
    leftMotor->setVelocity(MAX_VELOCITY * .1);
    rightMotor->setVelocity(MAX_VELOCITY * -.1);
}

// Function that checks if the robot is in the process of moving forward.
// true -> still moving forward
// false -> robot has finished
bool ifForward() {

	// If the current encoder value is within .1 radians of the target encoder value
    if (abs(leftEnc->getValue() - targetEnc) < .1) {

		// Stop the robot and return false
        leftMotor->setVelocity(0);
        rightMotor->setVelocity(0);
        return false;
    }

	// Otherwise, return true
    return true;
}

// Function that checks if the robot is in the process of turning 90 degrees
// true -> still turning
// false -> robot has finished
bool ifTurn() {

	// If the current angle is more than 86 degrees greater than the initial angle, then stop
	// Note that the robot is set to stop at 86 degrees because the robot continues to move slightly
	// even after the motors are set to 0, making it stop approximately at 90 degrees. This can be improved with PID.
    if (abs(getAngle() - initAngle) >= 86.0) {
		
		// Stop the robot and return false
        leftMotor->setVelocity(0);
        rightMotor->setVelocity(0);
        return false;
    }

	// Otherwise, return true
    return true;
}


int main() {
    /*
     * There are two parts to the main program
     *      - initialization - we initialize the simulation, the robot, and its sensors
     *      - loop - here is where the logic goes to perform actual work.
     */


    /* Initialization */
    // setPosition tells the motor to run until it gets to the position we pass to it.  
    // Passing INFINITY sets the motors to velocity control mode, where the motors will continue running.
    leftMotor->setPosition(INFINITY);
    rightMotor->setPosition(INFINITY);

    // setVelocity sets the speed in rad/s.  0 means stop.
    leftMotor->setVelocity(0);
    rightMotor->setVelocity(0);

    // Retrieves the basic world time step and stores it in variable "timeStep". 
    int timeStep = (int)robot->getBasicTimeStep();

    // we enable the gyro sensor and the encoders
    // we need to pass the timeStep as the refresh rate of the sensor
    gyro->enable(timeStep);
    leftEnc->enable(timeStep);
    rightEnc->enable(timeStep);

    /* End of Initialization */

    /* Loop - where we do work*/

    // begin moving forward 2 meters 
    goForward(2.0);

	/*
	 * This is the main loop of the program, where the robot calls step() and passes in the timeStep.
	 * This loop should be periodically called. Failing to call step() for an extended period of time 
	 * will break the controller. Motors and sensors only update following a step() call. However, you
	 * should only be calling step() at one place in your program, inside of the condition of your main loop.
	 */
    while (robot->step(timeStep) != -1) {

        // Update the angle heading by integrating the gyro's current reading
        updateGyro(timeStep);

        // If the robot state is set to moving forward, and ifForward() tells us that the robot has finished
		// moving forward, then transition into turning 90 degrees.
        if (state == movingForward && !ifForward()) {

			// Switch to turning
            doTurn();
        }
		// Otherwise, check if the state is set to turning and ifTurn() tells us that the robot has finished turning.
        else if (state == turn && !ifTurn()) {

			// If numEdgesTravel is less than 3, meaning that we have not gone fully around in a box yet, then
			// increment numEdgesTravel and start moving forward 2 meters again.
            if (numEdgesTravel < 3) {

                numEdgesTravel++;
                goForward(2.0);
            }
        }
    }

    return 0;
}